### PR TITLE
[PyTorch Numeric Suite] Add module level comparison

### PR DESCRIPTION
**Pull request by hx89**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #36701 [PyTorch Numeric Suite] Add module output comparison
* **#36669 [PyTorch Numeric Suite] Add module level comparison**

Add module level comparison API.

Differential Revision: [D21045393](https://our.internmc.facebook.com/intern/diff/D21045393/)

### CODE REVIEW

File: test/quantization/test_numeric_suite.py [hunk 1]
```diff
@@ -22,3 +26,49 @@ def test_compare_weights(self):
         self.assertEqual(len(weight_dict), 1)
         for k, v in weight_dict.items():
             self.assertTrue(v["float"].shape == v["quantized"].shape)
+
```
Comment (raghuramank100) [inline test/quantization/test_numeric_suite.py:29]: Will be good to write this as a loop to reduce code:
``` python
 def test_compare_weights(self):
      model_list = [AnnotatedConvModel(), AnnotatedConvBnReLUModel()]
      for model in model_list:
             model.eval()
             ...
```

File: test/quantization/test_numeric_suite.py [hunk 2]
```diff
@@ -22,3 +26,49 @@ def test_compare_weights(self):
...
+    def test_compare_model_stub(self):
+        r"""Compare the output of quantized conv layer and its float shadow module
+        """
+        # eager mode
+        annotated_conv_model = AnnotatedConvModel().eval()
+        quantized_annotated_conv_model = quantize(
+            annotated_conv_model, default_eval_fn, self.img_data
+        )
+        data = self.img_data[0][0]
+        module_swap_list = [torch.nn.Conv2d]
```
Comment (raghuramank100) [inline test/quantization/test_numeric_suite.py:52]: Similar comments as above. If you make module_swap_list = [ torch.nn.Conv2d, nn.intrinsic.modules.fused.ConvReLU2d] , it should work.

File: test/quantization/test_numeric_suite.py [hunk 3]
```diff
@@ -22,3 +26,49 @@ def test_compare_weights(self):
...
+        for k, v in ob_dict.items():
+            self.assertTrue(v["float"].shape == v["quantized"].shape)
+
+        annotated_conv_bn_relu_model = AnnotatedConvBnReLUModel().eval()
+        annotated_conv_bn_relu_model.fuse_model()
+        quantized_annotated_conv_bn_relu_model = quantize(
+            annotated_conv_bn_relu_model, default_eval_fn, self.img_data
+        )
+        module_swap_list = [nn.intrinsic.modules.fused.ConvReLU2d]
+        ob_dict = compare_model_stub(
```
Comment (raghuramank100) [inline test/quantization/test_numeric_suite.py:66]: Please add an additional test for swapping a user defined module here. For example
```python
class ModelWithSubModules(torch.nn.Module):
    def __init__(self):
        super(ModelWithSubModules, self).__init__()
        self.qconfig = default_qconfig
        self.mod1 = subModule()
        self.conv = torch.nn.Conv2d(...)
        self.quant = QuantStub()
        self.dequant = DeQuantStub()

def forward(self, x):
     x = self.quant(x)
     x = self.mod1(x)
     x = self.conv(x)
     x= self.dequant(x)
    return x
```
 The test should add a stub only to mod1

File: torch/quantization/_numeric_suite.py [hunk 4]
```diff
@@ -39,3 +42,139 @@ def compare_weights(float_dict, quantized_dict):
...
+            used to compare the module outputs between float and quantized
+            models
+        target_dict: the dictionary used to save all the observers
+    """
+
+    def get_prefix(prefix):
+        return prefix if prefix == "" else prefix + "."
+
+    if hasattr(mod, observer_name):
+        if observer_name == "logger":
```
Comment (raghuramank100) [inline torch/quantization/_numeric_suite.py:65]: Instead of matching strings, suggest having a baseclass for logger and for tensor observers. As long as someone inherits this base-class, we can match.
i.e do
 ```
 if type(observer) == logger:
  ```

File: torch/quantization/_numeric_suite.py [hunk 5]
```diff
@@ -39,3 +42,139 @@ def compare_weights(float_dict, quantized_dict):
             weight_dict[key]["float"] = float_dict[match_key]
             weight_dict[key]["quantized"] = quantized_dict[key]
     return weight_dict
+
+
+def _get_observer_dict(mod, target_dict, observer_name, prefix=""):
```
Comment (raghuramank100) [inline torch/quantization/_numeric_suite.py:47]: Suggest making this a top level API, i.e remove the _ Also, instead of observer name, lets go with observer type.

File: torch/quantization/_numeric_suite.py [hunk 6]
```diff
@@ -39,3 +42,139 @@ def compare_weights(float_dict, quantized_dict):
...
+            target_dict[
+                get_prefix(prefix) + "stats"
+            ] = mod.activation_post_process.tensor_val
+
+    for name, child in mod.named_children():
+        module_prefix = get_prefix(prefix) + name if prefix else name
+        _get_observer_dict(child, target_dict, observer_name, module_prefix)
+
+
+class Logger(nn.Module):
```
Comment (raghuramank100) [inline torch/quantization/_numeric_suite.py:77]: Create a base class that only has a stats dict but no forward method and inherit logger from that class.

File: torch/quantization/_numeric_suite.py [hunk 7]
```diff
@@ -39,3 +42,139 @@ def compare_weights(float_dict, quantized_dict):
...
+            self.stats["float"] = torch.cat((self.stats["float"], y.detach()))
+
+
+class Shadow(nn.Module):
+    r"""Shadow module attaches the float module to its matching quantized module
+    as the shadow. Then it uses Logger module to process the outputs of both
+    modules to do the comparison.
+    """
+
+    def __init__(self, q_module, float_module, Logger):
```
Comment (raghuramank100) [inline torch/quantization/_numeric_suite.py:115]: We need to add more documentation to explain the arguments for the init function for the class.

File: torch/quantization/_numeric_suite.py [hunk 8]
```diff
@@ -39,3 +42,139 @@ def compare_weights(float_dict, quantized_dict):
...
+
+    def forward(self, x):
+        output = self.orig_module(x)
+        x = x.dequantize()
+        shadow_output = self.shadow_module(x)
+        self.logger(output, shadow_output)
+        return output
+
+
+def _prepare_model_with_stubs(float_module, q_module, module_swap_list, Logger):
```
Comment (raghuramank100) [inline torch/quantization/_numeric_suite.py:130]: Lets make this a top level function, the reason is that users can insert arbitrary code between preparing and collecting stats. Perhaps they want to run several batches of data etc

File: torch/quantization/_numeric_suite.py [hunk 9]
```diff
@@ -39,3 +42,139 @@ def compare_weights(float_dict, quantized_dict):
...
+            _prepare_model_with_stubs(float_mod, mod, module_swap_list, Logger)
+
+        if type(float_mod) in module_swap_list:
+            reassign[name] = Shadow(mod, float_mod, Logger)
+
+    for key, value in reassign.items():
+        q_module._modules[key] = value
+
+
+def compare_model_stub(float_model, q_model, module_swap_list, data, Logger=Logger):
```
Comment (raghuramank100) [inline torch/quantization/_numeric_suite.py:163]: Wondering if we need this as a top level function or keep this as an internal/test only function and move it to the test code.

File: test/quantization/test_numeric_suite.py [hunk 10]
```diff
@@ -22,3 +26,49 @@ def test_compare_weights(self):
...
+            annotated_conv_bn_relu_model,
+            quantized_annotated_conv_bn_relu_model,
+            module_swap_list,
+            data,
+        )
+        self.assertEqual(len(ob_dict), 1)
+        for k, v in ob_dict.items():
```
Comment (raghuramank100) [inline test/quantization/test_numeric_suite.py:73]: Also, please add a test for functional modules (add and concat) to ensure that the API works for that also.

**Review by raghuramank100 — CHANGES_REQUESTED** [review]
Great start!.  Few suggestions for further improvement

File: torch/quantization/_numeric_suite.py (hunk 9 above)
Comment (hx89) [reply torch/quantization/_numeric_suite.py:163]: Maybe we can keep it as top level function since it's easy for user to use as a one-liner to quickly check the model accuracy?

**Comment by hx89** [comment]
Address comments.

File: torch/quantization/_numeric_suite.py [hunk 11]
```diff
@@ -39,3 +43,203 @@ def compare_weights(float_dict, quantized_dict):
...
+    shadow modules
+    """
+
+    def __init__(self):
+        super(RecordingLogger, self).__init__()
+        self.stats["float"] = None
+        self.stats["quantized"] = None
+
+    def forward(self, x, y):
+        print("in RecordingLogger forward")
```
Comment (raghuramank100) [inline torch/quantization/_numeric_suite.py:98]: Please remove print

File: test/quantization/test_numeric_suite.py [hunk 12]
```diff
@@ -1,24 +1,155 @@
...
+        self.assertTrue(isinstance(q_model.mod1, Shadow))
+        self.assertFalse(isinstance(q_model.conv, Shadow))
+        for k, v in ob_dict.items():
+            torch.testing.assert_allclose(v["float"], v["quantized"])
+
+        # Test adding stub to functionals
+        model = ModelWithFunctionals().eval()
+        q_model = copy.deepcopy(model)
+        q_model.qconfig = torch.quantization.get_default_qconfig("fbgemm")
+        q_model = prepare(q_model)
```
Comment (raghuramank100) [inline test/quantization/test_numeric_suite.py:140]: You can remove the deepcopy by calling prepare with inplace=False

File: torch/quantization/_numeric_suite.py (hunk 9 above)
Comment (raghuramank100) [reply torch/quantization/_numeric_suite.py:228]: Sure, as long as we have the building blocks as user facing functions, we are ok.

**Review by raghuramank100 — APPROVED** [review]
Looks good, have a few comments
